File: dropper_trap/trap.js
// ============================================================================
// DROPPER TRAP v4 — JavaScript Hooks (BEHAVIORAL + OPTIMIZED)
// ============================================================================
// Changes from v3:
//   - BEHAVIORAL: any fs.write* / fs.append* targeting monitor/resource/
//     (cl_playerlist|sv_main|sv_resources).lua from a non-monitor resource
//     is blocked regardless of content. Catches txAdmin tampering even when
//     the family rotates marker strings.
//   - REPORTING: optional one-line console banner at startup with the issue
//     submission URL. No automatic reporting; copy-paste only.
//
// IOC SOURCE OF TRUTH: iocs/blum_iocs.json in this repo. The lists below are
// a runtime mirror; when updating, edit the JSON first and mirror here.
// ============================================================================

const fs = require('fs');
const path = require('path');

const orig = {
    writeFile: fs.writeFile,
    writeFileSync: fs.writeFileSync,
    appendFile: fs.appendFile,
    appendFileSync: fs.appendFileSync,
    readFile: fs.readFile,  // keep clean ref for our own scans
};

const origSRF = typeof SaveResourceFile === 'function' ? SaveResourceFile : null;

const SUSPICIOUS = [
    'String.fromCharCode', 'fromCharCode', 'bertjj', 'bertJJ',
    'miauss', 'miausas', 'fivems.lt', '9ns1.com',
    'blum-panel', 'warden-panel', 'cipher-panel', 'gfxpanel',
    'RESOURCE_EXCLUDE', 'isExcludedResource', 'onServerResourceFail',
    'decompressFromUTF16', '\u15E1', 'ggWP', 'helpEmptyCode',
    'JohnsUrUncle', 'txadmin:js_create',
];

const TARGETS = new Set([
    'yarn_builder.js', 'webpack_builder.js',
    'sv_main.lua', 'sv_resources.lua', 'main.js', 'script.js',
    'babel_config.js', 'jest_mock.js', 'mock_data.js', 'commands.js', 'cl_playerlist.lua',
]);

let blockedCount = 0;

function isSuspicious(data) {
    if (!data) return null;
    // Only check first 2KB — malware signatures are always near the top
    let str;
    if (typeof data === 'string') {
        str = data.length > 2048 ? data.substring(0, 2048) : data;
    } else if (Buffer.isBuffer(data)) {
        str = data.toString('utf8', 0, Math.min(data.length, 2048));
    } else {
        return null;
    }
    for (const p of SUSPICIOUS) { if (str.includes(p)) return p; }
    return null;
}

function isTarget(fp) {
    if (!fp) return false;
    return TARGETS.has(path.basename(fp.toString()));
}

// Behavioral guard: writes to txAdmin monitor files from any non-monitor
// resource are tampering regardless of content.
const PROTECTED_TXADMIN = new Set(['cl_playerlist.lua', 'sv_main.lua', 'sv_resources.lua']);
function isProtectedTxAdminPath(fp) {
    if (!fp) return false;
    const s = fp.toString();
    if (!s.toLowerCase().includes('monitor')) return false;
    return PROTECTED_TXADMIN.has(path.basename(s).toLowerCase());
}

function getRes() {
    try { return GetInvokingResource() || GetCurrentResourceName() || '?'; }
    catch(e) { return '?'; }
}


// ============================================================================
// HOOK: fs.writeFile / writeFileSync — only check target files
// ============================================================================
fs.writeFile = function(filepath, data, ...args) {
    // BEHAVIORAL: any write to a txAdmin monitor file from a non-monitor
    // resource is blocked regardless of content. Defends against marker rotation.
    if (isProtectedTxAdminPath(filepath)) {
        const invoker = getRes();
        if (invoker !== 'monitor' && invoker !== 'dropper_trap') {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED txAdmin tamper: ${filepath} written by ${invoker} (behavioral)^0`);
            const cb = args.find(a => typeof a === 'function');
            if (cb) cb(null);
            return;
        }
    }
    if (isTarget(filepath)) {
        const match = isSuspicious(data);
        if (match) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED writeFile: ${filepath} | ${match} | ${getRes()}^0`);
            const cb = args.find(a => typeof a === 'function');
            if (cb) cb(null);
            return;
        }
    }
    return orig.writeFile.call(fs, filepath, data, ...args);
};

fs.writeFileSync = function(filepath, data, ...args) {
    if (isProtectedTxAdminPath(filepath)) {
        const invoker = getRes();
        if (invoker !== 'monitor' && invoker !== 'dropper_trap') {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED txAdmin tamper: ${filepath} written by ${invoker} (behavioral)^0`);
            return;
        }
    }
    if (isTarget(filepath)) {
        const match = isSuspicious(data);
        if (match) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED writeFileSync: ${filepath} | ${match} | ${getRes()}^0`);
            return;
        }
    }
    return orig.writeFileSync.call(fs, filepath, data, ...args);
};


// ============================================================================
// HOOK: fs.appendFile / appendFileSync — only check target files
// ============================================================================
fs.appendFile = function(filepath, data, ...args) {
    if (isProtectedTxAdminPath(filepath)) {
        const invoker = getRes();
        if (invoker !== 'monitor' && invoker !== 'dropper_trap') {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED txAdmin tamper: ${filepath} appended by ${invoker} (behavioral)^0`);
            const cb = args.find(a => typeof a === 'function');
            if (cb) cb(null);
            return;
        }
    }
    if (isTarget(filepath)) {
        const match = isSuspicious(data);
        if (match) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED appendFile: ${filepath} | ${match} | ${getRes()}^0`);
            const cb = args.find(a => typeof a === 'function');
            if (cb) cb(null);
            return;
        }
    }
    return orig.appendFile.call(fs, filepath, data, ...args);
};

fs.appendFileSync = function(filepath, data, ...args) {
    if (isProtectedTxAdminPath(filepath)) {
        const invoker = getRes();
        if (invoker !== 'monitor' && invoker !== 'dropper_trap') {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED txAdmin tamper: ${filepath} appended by ${invoker} (behavioral)^0`);
            return;
        }
    }
    if (isTarget(filepath)) {
        const match = isSuspicious(data);
        if (match) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED appendFileSync: ${filepath} | ${match} | ${getRes()}^0`);
            return;
        }
    }
    return orig.appendFileSync.call(fs, filepath, data, ...args);
};


// ============================================================================
// HOOK: SaveResourceFile — BLOCK backdoor content
// ============================================================================
if (origSRF) {
    global.SaveResourceFile = function(resourceName, fileName, data, dataLength) {
        const match = isSuspicious(data);
        if (match) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED SaveResourceFile: ${resourceName}/${fileName} | ${match} | ${getRes()}^0`);
            return false;
        }
        return origSRF(resourceName, fileName, data, dataLength);
    };
}


// ============================================================================
// HOOK: HTTPS — Block connections to known C2 domains
// ============================================================================
try {
    const https = require('https');
    const origGet = https.get;
    const origReq = https.request;

    const C2 = new Set([
        '9ns1.com','fivems.lt','blum-panel.me','blum-panel.com',
        'warden-panel.me','jking.lt','gfxpanel.org',
        '0xchitado.com','giithub.net','fivemgtax.com',
        'bhlool.com','flowleakz.org','z1lly.org',
        'l00x.org','monloox.com','ryenz.net','spacedev.fr',
        'noanimeisgay.com','trezz.org','2ns3.net','5mscripts.net',
        'kutingplays.com','bybonvieux.com','iwantaticket.org',
        '2312321321321213.com','2nit32.com','useer.it.com','wsichkidolu.com',
        'cipher-panel.me','ciphercheats.com','keyx.club','dark-utilities.xyz',
    ]);

    function isC2(url) {
        const s = typeof url === 'string' ? url : (url && (url.hostname || url.host || ''));
        if (!s) return false;
        for (const d of C2) { if (s.includes(d)) return d; }
        return false;
    }

    // Shared fake request object — reuse instead of creating EventEmitters
    function makeFake() {
        const noop = () => fake;
        const fake = { on: noop, end: noop, destroy: noop, write: noop, setTimeout: noop, once: noop, addListener: noop };
        return fake;
    }

    https.get = function(url, ...args) {
        const domain = isC2(url);
        if (domain) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED C2: ${domain} | ${getRes()}^0`);
            return makeFake();
        }
        return origGet.call(https, url, ...args);
    };

    https.request = function(url, ...args) {
        const domain = isC2(url);
        if (domain) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED C2: ${domain} | ${getRes()}^0`);
            return makeFake();
        }
        return origReq.call(https, url, ...args);
    };
} catch(e) {}


// ============================================================================
// HOOK: eval — BLOCK if suspicious
// ============================================================================
const origEval = global.eval;
global.eval = function(code) {
    if (typeof code === 'string') {
        const match = isSuspicious(code);
        if (match) {
            blockedCount++;
            console.log(`^1[TRAP-JS] BLOCKED eval | ${match} | ${getRes()}^0`);
            return undefined;
        }
    }
    return origEval(code);
};


// ============================================================================
// PERIODIC: Mutex check — every 30s
// ============================================================================
setInterval(() => {
    try {
        const gs = typeof GlobalState !== 'undefined' ? GlobalState : null;
        if (!gs) return;
        for (const name of ['miauss', 'miausas', 'ggWP']) {
            const val = gs[name];
            if (val != null) {
                console.log(`^1[TRAP-JS] MUTEX: GlobalState.${name} = "${val}" — CLEARING^0`);
                gs[name] = null;
            }
        }
    } catch(e) {}
}, 30000);


// ============================================================================
// PERIODIC: File scan — every 120s, ASYNC reads (non-blocking)
// ============================================================================
setInterval(() => {
    try {
        const numRes = GetNumResources();
        let found = 0;
        let pending = 0;

        for (let i = 0; i < numRes; i++) {
            const resName = GetResourceByFindIndex(i);
            if (!resName || resName === 'dropper_trap') continue;

            let resPath;
            try { resPath = GetResourcePath(resName); } catch(e) { continue; }
            if (!resPath) continue;

            for (const target of TARGETS) {
                pending++;
                // ASYNC read — does NOT block server thread
                orig.readFile.call(fs, path.join(resPath, target), 'utf8', (err, content) => {
                    pending--;
                    if (!err && content) {
                        const match = isSuspicious(content);
                        if (match) {
                            found++;
                            console.log(`^1[TRAP-JS] INFECTED: ${resName}/${target} | ${match}^0`);
                        }
                    }
                });
            }
        }
    } catch(e) {}
}, 120000);


console.log('^2[TRAP-JS] v4 ACTIVE | hooks: fs.write(behavioral txAdmin block) SaveResourceFile https eval | scan: 120s async^0');
console.log('^2[TRAP-JS] To report unrecognised blocks (optional): https://github.com/ImJer/blum-panel-fivem-backdoor-analysis/issues/new?template=scanner-findings.md  (no auto-reporting)^0');

File: dropper_trap/trap.lua
-- ============================================================================
-- DROPPER TRAP v4 — Lua Hooks (BEHAVIORAL + OPTIMIZED)
-- ============================================================================
-- Changes from v3:
--   - BEHAVIORAL: any write to monitor/resource/(cl_playerlist|sv_main|
--     sv_resources).lua from a non-monitor resource is blocked regardless
--     of content. Catches txAdmin tampering even when markers are renamed.
--   - SHADOW: pre-registers known backdoor net events (onServerResourceFail,
--     txadmin:js_create, helpEmptyCode) so the malicious handlers can't run.
--   - MANIFEST WATCHER: hashes every fxmanifest.lua at first scan; alerts on
--     runtime changes (catches manifest tampering regardless of content).
--   - REPORTING: optional one-line console banner at startup with the issue
--     submission URL. No automatic reporting; copy-paste only.
--
-- IOC SOURCE OF TRUTH: iocs/blum_iocs.json in this repo. The lists below are
-- a runtime mirror; when updating, edit the JSON first and mirror here.
-- ============================================================================

local SUSPICIOUS_PATTERNS = {
    "String.fromCharCode",
    "fromCharCode",
    "bertjj", "bertJJ",
    "miauss", "miausas",
    "fivems.lt",
    "9ns1.com",
    "blum-panel",
    "warden-panel",
    "cipher-panel",
    "gfxpanel",
    "RESOURCE_EXCLUDE",
    "isExcludedResource",
    "onServerResourceFail",
    "decompressFromUTF16",
    "\\u15E1",
    "ggWP",
    "helpEmptyCode",
    "JohnsUrUncle",
    "txadmin:js_create",
}

local KNOWN_TARGETS = {
    "yarn_builder.js", "webpack_builder.js",
    "sv_main.lua", "sv_resources.lua",
    "main.js", "script.js",
    "babel_config.js", "jest_mock.js",
    "mock_data.js", "commands.js",
    "cl_playerlist.lua",
}

-- O(1) lookup
local TARGET_SET = {}
for _, t in ipairs(KNOWN_TARGETS) do TARGET_SET[t] = true end

local blockedCount = 0

local function isSuspicious(content)
    if not content or type(content) ~= "string" then return nil end
    for _, pattern in ipairs(SUSPICIOUS_PATTERNS) do
        if content:find(pattern, 1, true) then return pattern end
    end
    return nil
end

local function isKnownTarget(filepath)
    if not filepath then return false end
    local name = filepath:match("([^/\\]+)$")
    return name and TARGET_SET[name] or false
end

-- Behavioral guard: writes to these txAdmin monitor files from a non-monitor
-- resource are tampering regardless of content. Catches the family even when
-- they rotate marker strings.
local PROTECTED_TXADMIN_FILES = {
    ["cl_playerlist.lua"] = true,
    ["sv_main.lua"] = true,
    ["sv_resources.lua"] = true,
}

local function isProtectedTxAdminPath(filepath)
    if not filepath then return false end
    local lower = filepath:lower()
    if not lower:find("monitor") then return false end
    local name = filepath:match("([^/\\]+)$")
    return name and PROTECTED_TXADMIN_FILES[name:lower()] or false
end

local function getResource()
    return GetInvokingResource() or GetCurrentResourceName() or "unknown"
end


-- ============================================================================
-- HOOK: io.open — ONLY wraps writes to known target files
-- Everything else passes through with ZERO overhead
-- ============================================================================
local origIoOpen = io.open
io.open = function(filepath, mode, ...)
    mode = mode or "r"

    -- Fast path: reads pass straight through
    if not (mode:find("w") or mode:find("a")) then
        return origIoOpen(filepath, mode, ...)
    end

    -- Fast path: not a target = pass through silently (NO LOGGING)
    if not isKnownTarget(filepath) then
        return origIoOpen(filepath, mode, ...)
    end

    -- Only reaches here for writes to known target files
    local realHandle = origIoOpen(filepath, mode, ...)
    if not realHandle then return nil end

    local fakeHandle = {}
    setmetatable(fakeHandle, {__index = realHandle})

    fakeHandle.write = function(self, data, ...)
        -- BEHAVIORAL: any write to a txAdmin monitor file from a resource
        -- other than 'monitor' is tampering regardless of content. This
        -- defends against marker rotation by Blum-family variants.
        if isProtectedTxAdminPath(filepath) then
            local invoker = getResource()
            if invoker ~= "monitor" and invoker ~= "dropper_trap" then
                blockedCount = blockedCount + 1
                print("^1[TRAP] BLOCKED txAdmin tamper: " .. tostring(filepath) ..
                      " written by " .. invoker .. " (behavioral; content not checked)^0")
                return
            end
        end
        if type(data) == "string" then
            local match = isSuspicious(data)
            if match then
                blockedCount = blockedCount + 1
                print("^1[TRAP] BLOCKED " .. tostring(filepath) .. " | " .. match .. " | " .. getResource() .. "^0")
                return
            end
        end
        return realHandle:write(data, ...)
    end

    fakeHandle.close = function(self)
        return realHandle:close()
    end

    return fakeHandle
end


-- ============================================================================
-- HOOK: os.execute + io.popen — ALWAYS BLOCK
-- ============================================================================
os.execute = function(command, ...)
    print("^1[TRAP] BLOCKED os.execute: " .. tostring(command):sub(1, 100) .. " | " .. getResource() .. "^0")
    return nil, "exit", 1
end

local origIoPopen = io.popen
if origIoPopen then
    io.popen = function(command, ...)
        print("^1[TRAP] BLOCKED io.popen: " .. tostring(command):sub(1, 100) .. " | " .. getResource() .. "^0")
        return nil, "blocked"
    end
end


-- ============================================================================
-- HOOK: load / loadstring — BLOCK backdoor patterns
-- ============================================================================
local origLoad = load
load = function(chunk, ...)
    if type(chunk) == "string" then
        local match = isSuspicious(chunk)
        if match then
            print("^1[TRAP] BLOCKED load() | " .. match .. " | " .. getResource() .. "^0")
            return function() end, nil
        end
    end
    return origLoad(chunk, ...)
end

local origLoadstring = loadstring
if origLoadstring then
    loadstring = function(chunk, ...)
        if type(chunk) == "string" then
            local match = isSuspicious(chunk)
            if match then
                print("^1[TRAP] BLOCKED loadstring() | " .. match .. " | " .. getResource() .. "^0")
                return function() end, nil
            end
        end
        return origLoadstring(chunk, ...)
    end
end


-- ============================================================================
-- HOOK: SaveResourceFile — BLOCK backdoor content
-- ============================================================================
local origSaveResourceFile = SaveResourceFile
if origSaveResourceFile then
    SaveResourceFile = function(resourceName, fileName, data, dataLength, ...)
        if isSuspicious(data) then
            blockedCount = blockedCount + 1
            print("^1[TRAP] BLOCKED SaveResourceFile: " .. resourceName .. "/" .. fileName .. " | " .. getResource() .. "^0")
            return false
        end
        return origSaveResourceFile(resourceName, fileName, data, dataLength, ...)
    end
end


-- ============================================================================
-- SHADOW REGISTER: known backdoor net events
-- We register these first so our handler runs before any malicious handler
-- and CancelEvent()s the propagation. IMPORTANT: dropper_trap must load
-- before any infected resource for this defense to work — put `ensure
-- dropper_trap` as the FIRST `ensure` line in your resources.cfg.
--
-- To add a new family event name: edit iocs/blum_iocs.json under
-- txadmin_tampering.backdoor_event_names, then mirror the change here.
-- ============================================================================
local SHADOW_EVENTS = {
    "onServerResourceFail",   -- Blum/Warden server-side RCE
    "txadmin:js_create",      -- Blum JS execution variant
    "helpEmptyCode",          -- defensive: in case the family ever flips it server-side
}

for _, eventName in ipairs(SHADOW_EVENTS) do
    RegisterNetEvent(eventName)
    AddEventHandler(eventName, function(...)
        blockedCount = blockedCount + 1
        print("^1[TRAP] BLOCKED backdoor event '" .. eventName ..
              "' from player " .. tostring(source) .. "^0")
        CancelEvent()
    end)
end


-- ============================================================================
-- PERIODIC: Mutex check — every 30s (just reads GlobalState, near-zero cost)
-- ============================================================================
CreateThread(function()
    while true do
        Wait(30000)
        for _, name in ipairs({"miauss", "miausas", "ggWP"}) do
            local val = GlobalState[name]
            if val ~= nil then
                print("^1[TRAP] MUTEX: GlobalState." .. name .. " = " .. tostring(val) .. " — CLEARING^0")
                GlobalState[name] = nil
            end
        end
    end
end)


-- ============================================================================
-- MANIFEST WATCHER: hash every fxmanifest at first scan; alert on changes.
-- Catches manifest tampering at runtime regardless of marker rotation. Blum
-- typically modifies fxmanifest.lua to inject hidden script paths; this
-- detects any change to manifest content after the first observation.
-- ============================================================================
local manifestHashes = {}

-- Simple non-cryptographic hash sufficient to detect any manifest change.
-- Lua doesn't ship sha2 in FiveM's runtime; collision resistance isn't
-- needed here because we only compare against the prior value of the same
-- file, not across files.
local function simpleHash(s)
    local h = 0
    for i = 1, #s do
        h = (h * 31 + s:byte(i)) % 2147483647
    end
    return h
end


-- ============================================================================
-- PERIODIC: File scan — every 120s, staggered with Wait(0) between resources
-- ============================================================================
CreateThread(function()
    Wait(30000)  -- let server finish booting
    while true do
        local numResources = GetNumResources()
        local found = 0
        for i = 0, numResources - 1 do
            local resName = GetResourceByFindIndex(i)
            if resName and resName ~= "dropper_trap" then
                for _, target in ipairs(KNOWN_TARGETS) do
                    local content = LoadResourceFile(resName, target)
                    if content and isSuspicious(content) then
                        found = found + 1
                        print("^1[TRAP] INFECTED: " .. resName .. "/" .. target .. "^0")
                    end
                end
                -- Manifest watcher: capture hash on first sight; alert on diff.
                local manifest = LoadResourceFile(resName, "fxmanifest.lua") or
                                 LoadResourceFile(resName, "__resource.lua")
                if manifest then
                    local h = simpleHash(manifest)
                    if manifestHashes[resName] == nil then
                        manifestHashes[resName] = h
                    elseif manifestHashes[resName] ~= h then
                        print("^1[TRAP] MANIFEST CHANGED at runtime: " .. resName ..
                              " — possible tampering, review fxmanifest.lua^0")
                        manifestHashes[resName] = h
                    end
                end
                Wait(0)  -- yield every resource to prevent hitch
            end
        end
        if found > 0 then
            print("^1[TRAP] Scan: " .. found .. " infected file(s)^0")
            print("^3[TRAP] To share findings (optional, no auto-reporting): https://github.com/ImJer/blum-panel-fivem-backdoor-analysis/issues/new?template=scanner-findings.md^0")
        end
        Wait(120000)
    end
end)


print("^2[TRAP] v4 ACTIVE | hooks: io.open(behavioral txAdmin block) os.execute io.popen load SaveResourceFile | shadow events: " .. tostring(#SHADOW_EVENTS) .. " | manifest watcher: ON | scan: 120s^0")
print("^2[TRAP] To report unrecognised blocks (optional): https://github.com/ImJer/blum-panel-fivem-backdoor-analysis/issues/new?template=scanner-findings.md  (no auto-reporting)^0")
